src/balancecommand.cpp

```diff
@@ -0,0 +1,180 @@
+/****************************************************************************************
+**
+** Copyright (C) 2015 Jolla Ltd.
+** Contact: Martin Grimme <martin.grimme@gmail.com>
+** All rights reserved.
+**
+** This file is part of the btrfs-balancer package.
+**
+** You may use this file under the terms of the GNU Lesser General
+** Public License version 2.1 as published by the Free Software Foundation
+** and appearing in the file license.lgpl included in the packaging
+** of this file.
+**
+** This library is free software; you can redistribute it and/or
+** modify it under the terms of the GNU Lesser General Public
+** License version 2.1 as published by the Free Software Foundation
+** and appearing in the file license.lgpl included in the packaging
+** of this file.
+**
+** This library is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+** Lesser General Public License for more details.
+**
+****************************************************************************************/
+
+
+#include "balancecommand.h"
+#include "btrfsbalancer.h"
+#include "dbusservice.h"
+
+#include <QDBusConnection>
+#include <QDBusInterface>
+#include <QDBusMessage>
+#include <QFile>
+
+#include <contextproperty.h>
+
+#include <iostream>
+
+namespace
+{
+const QString BATTERY_CHARGE("Battery.ChargePercentage");
+const QString BATTERY_STATUS("Battery.IsCharging");
+
+int batteryCharge()
+{
+    return ContextProperty(BATTERY_CHARGE).value(QVariant(0)).toInt();
```

Comment:
I doubt you will get the right value with sync access like ^.

- ContextProperty process properties i/o in the separate thread to avoid blocking operations in the main thread and to improve performance (main event loop gets a lot of events related to graphics, so polls handles thousands (?) times/s). 

- Also, balancing operation could be power-consuming and if user takes device out when it is cold, battery level can drop very fast.

So, I recommend you to put into BallanceCommand class Battery.State and .ChargePercentage and connect to their valueChanged(). If State is "unknown" you can't be sure that ChargePercentage is correct. If it is "charging" or "full", you know what do you want, if discharging, track ChargePercentage.

P.S. I even recommend to track ChargePercentage when device is charging: if charger is weak and power consumption is high device could discharge.